contrib/golang/filters/http/source/go/pkg/http/filtermanager.go

```diff
@@ -63,3 +81,271 @@ func getHttpFilterConfigParser(name string) api.StreamFilterConfigParser {
+	conf, ok := c.(*filterManagerConfig)
+	if !ok {
+		panic("unexpected config type")
+	}
+
+	newConfig := conf.current
+	factories := make([]api.StreamManagedFilterFactory, len(newConfig))
+	names := make([]string, len(factories))
+	for i, fc := range newConfig {
+		var factory api.StreamManagedFilterConfigFactory
+		name := fc.Name
+		names[i] = name
+		if v, ok := httpManagedFilterConfigFactoryAndParser.Load(name); ok {
+			plugin := v.(*managedFilterConfigFactoryAndParser)
+			factory = plugin.configFactory
+			config := fc.parsedConfig
+			factories[i] = factory(config)
+
+		} else {
+			api.LogErrorf("plugin %s not found, pass through by default", name)
+			factory = ManagedPassThroughFactory
+			factories[i] = factory(nil)
+		}
+	}
+
+	return func(callbacks api.FilterCallbackHandler) api.StreamFilter {
+		filters := make([]api.StreamFilter, len(factories))
+		for i, factory := range factories {
+			filters[i] = factory(callbacks)
+		}
+		return &filterManager{
+			callbacks: callbacks,
+			names:     names,
+			filters:   filters,
+		}
+	}
+}
+
+func (m *filterManager) DecodeHeaders(header api.RequestHeaderMap, endStream bool) api.StatusType {
+	go func() {
+		defer m.callbacks.RecoverPanic()
+
+		for _, f := range m.filters {
+			// When the filter is controlled by filterManager, it don't
+			// need to create goroutine by itself. So there is no need
+			// to return api.Running and resume inside the filter. We
+			// treat api.Running as api.Continue here.
+			status := f.DecodeHeaders(header, endStream)
+			if status == api.LocalReply {
+				return
+			}
+		}
+		m.callbacks.Continue(api.Continue)
+	}()
+
+	return api.Running
+}
+
+func (m *filterManager) notifyUnsupportedStatus(name string, status api.StatusType) (willReturn bool) {
+	if status == api.Running || status == api.Continue {
+		return false
+	}
+
+	// status == StopXXX
+	if len(m.filters) == 1 {
+		// just use the status as the final result
+		m.callbacks.Continue(status)
+		return true
+	}
+	// If we have multiple filters, we can't implement the StopXXX semantics without
+	// managing the buffer at Go's side. For example, considering we have two filters:
+	// - Compress (return Continue to process streamingly)
+	// - Rewrite (return StopAndBuffer to process the whole data)
+	//
+	// If we return StopAndBuffer as final result, the next time filter Compress will get the
+	// whole data, which is unwanted.
+	// However, managing the buffer at Go's side is inefficient, which will double the memory
+	// usage at least.
+
+	m.callbacks.Log(api.Error,
+		fmt.Sprintf("unsupported status %s returned from filter %s, ignored", status.String(), name))
+	return false
+}
+
+func (m *filterManager) DecodeData(buf api.BufferInstance, endStream bool) api.StatusType {
+	go func() {
+		defer m.callbacks.RecoverPanic()
+
+		for i, f := range m.filters {
+			status := f.DecodeData(buf, endStream)
+			if status == api.LocalReply {
```

Comment:
what if status = api.Running? it should be disabled, right?
I prefer to provide a new filter interface, with limited API statuses.